### ISSUE TITLE
Identity: ID vs. URL vs. URN

### ISSUE THREAD

**Issue by whiskeysierra** [zalando/restful-api-guidelines#62]
> favor simple IDs if in same namespace/system otherwise think about using URNs e.g. urn:warehouse:FG-AS-123-FD

relates to #54

**Comment by whiskeysierra** [comment]
See slides:
- [Identity of Resources](https://docs.google.com/a/zalando.de/presentation/d/1fUiAXClrQC56EOF4mIKkuZVezVAJT6H77VRk-8SsiSc/edit?usp=sharing)
- [ZRNs](https://docs.google.com/a/zalando.de/presentation/d/17ijwoEdLXxuxCPe9AXuMRxfk44JInBkWbf_s8r_Dlw8/edit?usp=sharing)

**Comment by tkrop** [comment]
As @whiskeysierra attached the slide I open the discussion.

I would suggest to provide an explicit recommendation to favor URIs over IDs if a system has no control over the ID spaces of the parameter. To foster the use of URIs e.g. in the form of ZRNs, we should propose that micro-services should provide either way to resolve an URI for an ID and make resources available under URIs. In effect we could suggest to make IDs obsolete over time by using URIs where IDs originally where used.

While URLs can be used in a similar way as URIs, they have the advantage/disadvantage of binding a location to the resource. If the location changes, or a resource can be handled in the context of multiple locations, this can result in problem, e.g. using an URL in an path parameter looks very strange, while an URI can be used consistently with high readability in any context.

**Comment by whiskeysierra** [comment]
```bnf
zrn ::= zrn:<system-id>:<resource-type>:<id>
system-id ::= [a-z0-9-.]*
resource-type ::= [a-z0-9-.]
identifier ::= .+
```

**Comment by whiskeysierra** [comment]
Maybe resource-type and id can be combined into one.

**Comment by tkrop** [comment]
I suggest to restrict the space for identifiers to `[a-z0-9-.]+` to allow un-encoded use of ZRNs in URL. The only time where I have seen other characters in IDs is the impractical case of URLs as IDs.

```bnf
zrn ::= zrn:<system-id>:<resource-type>:<id>
system-id ::= [a-z0-9-.]+
resource-type ::= [a-z0-9-.]+
identifier ::= [a-z0-9-.]+
```

**Comment by whiskeysierra** [comment]
What about `:`?

**Comment by jhorstmann** [comment]
@whiskeysierra can you summarize again the exact problem that zrn is supposed to solve?

**Comment by whiskeysierra** [comment]
It should solve the problem of distributed identities, i.e. systems that have conflicting id spaces. As an example you're offering *Fulfillment as a Service* and one of your input parameters is a sales order number, something that uniquely identifies something in another system. If you're dealing with many different systems those IDs are not unique anymore, so you need namespaces for them (`zrn:<system-id>`).

**Comment by jhorstmann** [comment]
I would argue that system-id and resource id should then be separate fields in apis. What's the benefit of encoding them into one string? I also think the resource type is nearly always known implicitly and there is no need to repeat it, like in `{"order_number": "zrn:zalando:order:1234"}`.

**Comment by whiskeysierra** [comment]
> I would argue that system-id and resource id should then be separate fields in apis. What's the benefit of encoding them into one string?

Self-contained, primitive value, I guess?

> I also think the resource type is nearly always known implicitly and there is no need to repeat it, like in {"order_number": "zrn:zalando:order:1234"}.

I'd expect this for a sales order:

```json
{
  "id": "zrn:zalando:sales-order:1234"
}
```

And something like this for a fulfillment order:

```json
{
  "order": "zrn:notebooksbilliger:bestellung:magento:1234"
}
```

Or `sales_order` or `sales_order_id` or `sales_order_ref`...?

**Comment by tkrop** [comment]
> What about :?

@whiskeysierra Possible, but why? Being restrictive has advantages when reading and validating ZRNs.

> I also think the resource type is nearly always known implicitly and there is no need to repeat it, like in `{"order_number": "zrn:zalando:order:1234"}`.

@jhorstmann this is implicitly known assumption is exactly what breaks you the neck when moving from an internal service to an public available service.

**Comment by jhorstmann** [comment]
> Self-contained, primitive value, I guess?

That's only a benefit if I never have to take this value apart. I would assume that different services are responsible for sales or fulfillment orders, so in order to get more information about the order I would have to look inside the zrn anyway. And if there would a facade that I could call with both types of zrn, I would assume that the response schema would be slightly different and I still need to know about the difference.

@tkrop Maybe my example was a bit too succinct. Would it still be too implicit if that snippet was returned from a service names "sales-order-service" or contained in an event like "sales-order-created"?

**Comment by tkrop** [comment]
@jhorstmann I think the point is, that `order_number` is now the wrong term. I should now be called `order_uri`. Besides, the important part is that all platform services should refer to all necessary externally referencable resources by their URIs, going so far that services should accept/expect URIs even in their path segments.

**Comment by jhorstmann** [comment]
Looking at [Willi's example for a fulfillment order](https://github.com/zalando/restful-api-guidelines/issues/62#issuecomment-243496348) again made it a bit clearer. I like that using urn instead of uri makes it clear that the value is not directly resolvable, and resolution has to go through some catalog (in the xml sense).

What would be required for this to work though, would be a central definition of all resource-ids inside the zalando namespace, to prevent multiple teams using the same terms. This would have to be more binding and strict than the current platform glossary.

**Comment by ePaul** [comment]
The point of using those resource names instead of URLs is when you don't need to resolve them, but just need it to compare/group other resources referring to them.

Yes, for this schema to work we need a registry of the "system ids" (i.e. "zalando", "notebooksbilliger" – though maybe we could use just a domain name here?), and for the zalando system we would need a registry of the resource types (something like identifier, short description, owning team/application ("who can generate those kinds of identifiers"), link to more documentation).

**Comment by tkrop** [comment]
@jhorstmann I known, this is an academic discussion, but URN is the outdated concept name for URI. Personally, I find URN(ame) less suitable than URI(dentifier), as the first does not express the unique constraint as the second.

@ePaul  even if we use URIs instead of URL to express that resources cannot be resolved directly, it would still be possible to have a protocol implementation for the schema "zrn:" that is able to resolve end-point information for the path segment and thus can directly resolve resources. In any case, we will have services that have to analyse the path segment to route subsequent requests. This services will require at least an internal registry that maps this information to service end-points for their purpose.

**Comment by whiskeysierra** [comment]
> Yes, for this schema to work we need a registry of the "system ids" (i.e. "zalando", "notebooksbilliger" – though maybe we could use just a domain name here?), and for the zalando system we would need a registry of the resource types (something like identifier, short description, owning team/application ("who can generate those kinds of identifiers"), link to more documentation).

In theory we have this with yourturn already, it's even called *Resource Types*. We only use it for OAuth right now, but that doesn't need to stay like that.

**Comment by whiskeysierra** [comment]
> resolution has to go through some catalog (in the xml sense).

I'm not sure we need this. Services should rely on all the necessary data being passed to them, rather than that they request it on their own, I'd say.

**Comment by whiskeysierra** [comment]
> I think the point is, that order_number is now the wrong term. I should now be called order_uri

That's like the hungarian notation. The property name should reflect the semantic, not the data type. You wouldn't rename `created_at` to `created_at_date_time`, would you? We use a URI as an identifier, so the field should be `order_id`.

**Comment by whiskeysierra** [comment]
If we'd use the `urn:<namespace>:<identifier>` scheme, we than wouldn't need to specify any constraints for our external systems (and we shouldn't care, basically). We could then focus on specifying ours:

    urn:com.zalando:sales-order:1234

Thoughts? I now for example that SAP is using URNs like that internally at some places.

**Comment by tkrop** [comment]
>>    I think the point is, that order_number is now the wrong term. I should now be called order_uri
>
> That's like the hungarian notation. The property name should reflect the semantic, not the data type. You  wouldn't rename created_at to created_at_date_time, would you? We use a URI as an identifier, so the field should be order_id.

@whiskeysierra no, the semantic of the field is definitely not an `order_number` and would thus be totally misleading. I have no problem to refer to a resource just by `order` because the remaining context is implicitly given by the context, but as long as I attach something that refers to the type, it should consequently be the real used type, e.g. `uri` or `ref` or even may be `id`/`uid`.

**Comment by whiskeysierra** [comment]
> no, the semantic of the field is definitely not an order_number and would thus be totally misleading.

My argument was not *in favor* or `order_number`, but rather *against* `order_uri`.

> even may be id/uid.

Exactly why my suggested was:

> We use a URI as an identifier, so the field should be order**_id**.

**Comment by jhorstmann** [comment]
> Services should rely on all the necessary data being passed to them, rather than that they request it on their own, I'd say.

This should also be part of the guidelines on its own n case it isn't already, or be part of the rationale for zrn.

**Comment by tkrop** [comment]
Hi, I created a [document describing the background](https://docs.google.com/document/d/1GdPK8Ooja0JL-k63wg1f2D2NTB-GCSd_7Ni-7NP_jpo/edit) to support the discussion.

Thanks to @whiskeysierra, @ePaul, @heidenwolf for giving valuable feedback.

**Comment by ePaul** [comment]
This will be tried out in some APIs. We are closing the issue for now, will be reopened for eventual guideline changes when we have more experience.
